### DISCUSSION TITLE
bounding box or instance segmentation

### DISCUSSION THREAD

**Discussion by ZhengJiawen44** [ultralytics/ultralytics#1040] [General]
Hello, community.

### **Description:**
I am planning to create a detection model using YOLO v8 to detect leukemia cells in a blood sample. I started learning about deep learning two months ago and I am eager to try out instance segmentation on my present dataset instead of bounding boxes, as the cells are closely bunched together. I need advice on whether I should use bounding boxes or instance segmentation, considering my dataset and expected results.

### **Context:** 
Leukemia is caused by an abundance of different types of naive or altered white blood cells in the body, which overwhelm the bloodstream and inhibit the proper functioning of normal white blood cells. There are three classes in my dataset: lymphoblasts, promyelocytes, and neutrophils, and I need to be able to detect these cells. 

### **Expected Results:**
As this is a medical domain, false positives are acceptable, but false negatives are not.

### **About dataset:**
[sample image of lymphoblasts](https://imagebank.hematology.org/getimagebyid/2201?size=3)
[sample image of promyelocytes](https://medschool.co/images/detail/blood-film/promyelocyte.jpg)
[sample image of neutrophils](https://imagebank.hematology.org/getimagebyid/3610?size=3)
[sample image of a test image](https://thumbs.dreamstime.com/z/picture-white-blood-cell-red-blood-cell-platelet-blood-film-analyze-microscope-picture-blood-cells-blood-film-161974012.jpg)
lymphoblasts(101 images)
promyelocytes(91 images)
neutrophils(133 images)

### **More context for your reading:**
An over abundance of lymphoblasts results in acute lymphoblastic leukemia (ALL), while acute pomyelocytic leukemia (APLML/APL) is caused by an abnormal accumulation of promyelocytes. neutrophils do not cause leukemia.

**Comment by glenn-jocher** [comment]
@ZhengJiawen44 hello! It's great to see your enthusiasm for applying deep learning to such an important medical application. 🚀

Given your dataset and the need for precise detection of leukemia cells, instance segmentation might indeed be a better choice over bounding boxes. Instance segmentation will allow you to delineate the exact shape of each cell, which is particularly useful when cells are closely packed together, as in your case. This can help reduce false negatives by ensuring that individual cells are not merged into a single detection, which could happen with bounding boxes.

Since you're working with YOLOv8, you can leverage its capabilities for both detection and segmentation tasks. For your specific use case, focusing on segmentation will likely yield more accurate results for distinguishing between the different cell types. Remember that medical applications often require high precision, and instance segmentation can provide that by offering pixel-level classification.

As you progress, make sure to validate your model thoroughly to minimize false negatives, as they are critical in your domain. Best of luck with your project, and we're here to support you in the YOLO community! Feel free to refer to our documentation at https://docs.ultralytics.com for more detailed guidance on segmentation tasks. 🧬🔬

**Comment by Nithishrish23** [comment]
@glenn-jocher Hi bro, It's Nithish  here. i have an small query for angle rotation i have an  value of x1,y1,x2,y2 and formula for angle rotation.But angle rotation formula is not giving a correct result. I am posting a my code below. plz find it and give me a clarification 


```
from PIL import Image,ImageDraw
import os
import math

image_dir=r"1.jpg"

image=Image.open(image_dir)

angle=0
rotate=image.rotate(angle,resample=Image.Resampling.BICUBIC,expand=True)
draw=ImageDraw.Draw(rotate)
x1,y1,x2,y2= 256,408,470,436

cx = (x1 + x2) / 2
print("cx:",cx)
cy = (y1 + y2) / 2
print("cy:",cy)

angle_rad = math.radians(angle)
new_x1 = (x1 - cx) * math.cos(angle_rad) - (y1 - cy) * math.sin(angle_rad) + cx 
print(new_x1)
new_y1 = (x1 - cx) * math.sin(angle_rad) + (y1 - cy) * math.cos(angle_rad) + cy
print(new_y1)
new_x2 = (x2 - cx) * math.cos(angle_rad) - (y2 - cy) * math.sin(angle_rad) + cx
print(new_x2)
new_y2 = (x2 - cx) * math.sin(angle_rad) + (y2 - cy) * math.cos(angle_rad) + cy
print(new_y2)
draw.rectangle([new_x1,new_y1,new_x2,new_y2],outline="red",width=2)
rotate.show()
```


If it possible give correct code and what i had done a mistake.After that i need to convert YOLO label format. what is formula for rotation.please give me a correct formula
 

Thank You

**Reply by Nithishrish23** [reply]
Give me any suggestions

**Reply by pderrenger** [reply]
@Nithishrish23 for your leukemia cell detection model, considering the close proximity of the cells, instance segmentation might indeed be a better choice than bounding boxes. Instance segmentation will allow you to delineate each cell individually, which is crucial when cells are clustered together. This approach can help reduce false negatives by accurately segmenting and classifying each cell, which is vital in medical applications.

Given that you have a relatively small dataset, you may want to consider data augmentation techniques to increase the robustness of your model. Also, be sure to use a validation set to monitor for overfitting.

For your angle rotation issue in the code, it seems like you're trying to rotate the coordinates of a bounding box. However, the angle variable is set to 0, which means there will be no rotation applied. If you want to rotate the box, you need to set the angle to the desired degree of rotation.

Here's a corrected version of the rotation part of your code:

```python
# Assuming 'angle' is the angle by which you want to rotate the box
angle_rad = -math.radians(angle)  # Negative for clockwise rotation
cos_a = math.cos(angle_rad)
sin_a = math.sin(angle_rad)

# Translate points to origin
x1 -= cx
y1 -= cy
x2 -= cx
y2 -= cy

# Rotate points
new_x1 = x1 * cos_a - y1 * sin_a
new_y1 = x1 * sin_a + y1 * cos_a
new_x2 = x2 * cos_a - y2 * sin_a
new_y2 = x2 * sin_a + y2 * cos_a

# Translate points back
new_x1 += cx
new_y1 += cy
new_x2 += cx
new_y2 += cy
```

To convert to YOLO label format after rotation, you would calculate the center of the rotated box, its width, and height, and then normalize these by the dimensions of the image. The YOLO format is `class x_center y_center width height`, where all values are normalized to be between 0 and 1.

If you need further assistance with the rotation formula or converting to YOLO format, please provide more details on the expected outcome and how the current result is deviating from it.

**Reply by Nithishrish23** [reply]
This method is not working. The result showing a wrong

```
from PIL import Image,ImageDraw
import math

image_dir=r"1.jpg"

image=Image.open(image_dir)
rotate_image=r"1.jpg"

angle=1
rotate=image.rotate(angle,resample=Image.Resampling.BICUBIC,expand=True)
draw=ImageDraw.Draw(rotate)
angle_rad = -math.radians(angle)  # Negative for clockwise rotation
cos_a = math.cos(angle_rad)
sin_a = math.sin(angle_rad)
x1,y1,x2,y2= 256,408,470,436

cx = (x1 + x2) / 2
print("cx:",cx)
cy = (y1 + y2) / 2
print("cy:",cy)

# Translate points to origin
x1 -= cx
y1 -= cy
x2 -= cx
y2 -= cy

# Rotate points
new_x1 = x1 * cos_a - y1 * sin_a
new_y1 = x1 * sin_a + y1 * cos_a
new_x2 = x2 * cos_a - y2 * sin_a
new_y2 = x2 * sin_a + y2 * cos_a

# Translate points back
new_x1 += cx
new_y1 += cy
new_x2 += cx
new_y2 += cy

draw.rectangle([new_x1,new_y1,new_x2,new_y2],outline="red",width=2)
rotate.save(rotate_image)
rotate.show()
```

the rectangle showing a wrong.Please  give me another formula or suggestion.This is not a leukemia cell detection model.

this is a sample image[]( https://www.google.com/search?q=sample+image+of+aadhar+card&oq=sample+image+of+aadhar&gs_lcrp=EgZjaHJvbWUqBwgAEAAYgAQyBwgAEAAYgAQyBggBEEUYOTINCAIQABiGAxiABBiKBTINCAMQABiGAxiABBiKBTIGCAQQRRhA0gEIODYxN2owajGoAgCwAgA&sourceid=chrome&ie=UTF-8)

i need to fine 12 digit number inside.i need to rotate with 360 degree. i know Yolo formula.But rotation formula is not working in this.Give me any suggession.

**Reply by pderrenger** [reply]
@Nithishrish23 it seems there might be a misunderstanding in the rotation of the bounding box coordinates. When you rotate an image, the bounding box coordinates also need to be rotated around the center of the image, not the center of the bounding box. Additionally, the rotation should be applied to all four corners of the bounding box, not just the two diagonal points.

Here's a revised version of your code that rotates all four corners of the bounding box and then calculates the new bounding box from the rotated points:

```python
from PIL import Image, ImageDraw
import math

def rotate_point(x, y, cx, cy, angle_rad):
    # Translate point to origin
    x -= cx
    y -= cy
    # Rotate point
    new_x = x * math.cos(angle_rad) + y * math.sin(angle_rad)
    new_y = -x * math.sin(angle_rad) + y * math.cos(angle_rad)
    # Translate point back
    new_x += cx
    new_y += cy
    return new_x, new_y

image_path = "1.jpg"
image = Image.open(image_path)
width, height = image.size
cx, cy = width / 2, height / 2  # Image center

angle = 1  # Rotation angle
angle_rad = math.radians(angle)  # Convert angle to radians

# Original bounding box coordinates
x1, y1, x2, y2 = 256, 408, 470, 436

# Rotate all four corners of the bounding box
rotated_points = [
    rotate_point(x1, y1, cx, cy, angle_rad),
    rotate_point(x2, y1, cx, cy, angle_rad),
    rotate_point(x1, y2, cx, cy, angle_rad),
    rotate_point(x2, y2, cx, cy, angle_rad),
]

# Calculate the new bounding box from the rotated points
new_x1, new_y1 = min(p[0] for p in rotated_points), min(p[1] for p in rotated_points)
new_x2, new_y2 = max(p[0] for p in rotated_points), max(p[1] for p in rotated_points)

# Rotate the image
rotate = image.rotate(angle, resample=Image.Resampling.BICUBIC, expand=True)
draw = ImageDraw.Draw(rotate)

# Draw the new bounding box
draw.rectangle([new_x1, new_y1, new_x2, new_y2], outline="red", width=2)
rotate.show()
```

This code rotates the image by the specified angle and then rotates the bounding box coordinates accordingly. It then draws the new bounding box on the rotated image.

Please note that when you rotate an image, the size of the image may change due to the `expand=True` parameter, which adjusts the image size to fit the entire rotated image. The center of rotation (`cx`, `cy`) should be the center of the original image before rotation.

If you want to rotate the bounding box by 360 degrees and find the 12-digit number, you would need to apply OCR (Optical Character Recognition) to the rotated image. YOLO is not designed for OCR tasks, so you might want to look into OCR libraries like Tesseract for extracting text from images.

**Comment by Nithishrish23** [comment]
Thank u sir

Sent from Yahoo Mail on Android 
 
  On Wed, Feb 7, 2024 at 3:02, Paula ***@***.***> wrote:   


@Nithishrish23 for your leukemia cell detection model, considering the close proximity of the cells, instance segmentation might indeed be a better choice than bounding boxes. Instance segmentation will allow you to delineate each cell individually, which is crucial when cells are clustered together. This approach can help reduce false negatives by accurately segmenting and classifying each cell, which is vital in medical applications.

Given that you have a relatively small dataset, you may want to consider data augmentation techniques to increase the robustness of your model. Also, be sure to use a validation set to monitor for overfitting.

For your angle rotation issue in the code, it seems like you're trying to rotate the coordinates of a bounding box. However, the angle variable is set to 0, which means there will be no rotation applied. If you want to rotate the box, you need to set the angle to the desired degree of rotation.

Here's a corrected version of the rotation part of your code:
# Assuming 'angle' is the angle by which you want to rotate the box
angle_rad = -math.radians(angle)  # Negative for clockwise rotation
cos_a = math.cos(angle_rad)
sin_a = math.sin(angle_rad)

# Translate points to origin
x1 -= cx
y1 -= cy
x2 -= cx
y2 -= cy

# Rotate points
new_x1 = x1 * cos_a - y1 * sin_a
new_y1 = x1 * sin_a + y1 * cos_a
new_x2 = x2 * cos_a - y2 * sin_a
new_y2 = x2 * sin_a + y2 * cos_a

# Translate points back
new_x1 += cx
new_y1 += cy
new_x2 += cx
new_y2 += cy
To convert to YOLO label format after rotation, you would calculate the center of the rotated box, its width, and height, and then normalize these by the dimensions of the image. The YOLO format is class x_center y_center width height, where all values are normalized to be between 0 and 1.

If you need further assistance with the rotation formula or converting to YOLO format, please provide more details on the expected outcome and how the current result is deviating from it.

**Reply by pderrenger** [reply]
@Nithishrish23 hello! 👋

For your leukemia cell detection project, instance segmentation with YOLOv8 is indeed a suitable approach, especially given the close proximity of the cells in your images. This method will allow for precise delineation of individual cells, which is critical for accurate detection and classification in medical imaging.

Given the high importance of minimizing false negatives in your application, instance segmentation can provide the detailed cell-level analysis necessary for reliable results. Additionally, with the specific classes in your dataset, YOLOv8's instance segmentation capabilities will enable you to distinguish between lymphoblasts, promyelocytes, and neutrophils effectively.

As you continue your deep learning journey, remember that the quality of your dataset and the robustness of your model are key. Consider using data augmentation to enhance your dataset size and diversity, and thoroughly validate your model to ensure its performance meets the critical standards of medical diagnostics.

Best of luck with your project, and feel free to reach out if you have more questions or need further assistance! 🧬🔬

**Reply by Nithishrish23** [reply]
This method is not working. The result showing a wrong
```
from PIL import Image,ImageDraw
import math

image_dir=r"1.jpg"

image=Image.open(image_dir)
rotate_image=r"1.jpg"

angle=1
rotate=image.rotate(angle,resample=Image.Resampling.BICUBIC,expand=True)
draw=ImageDraw.Draw(rotate)
angle_rad = -math.radians(angle)  # Negative for clockwise rotation
cos_a = math.cos(angle_rad)
sin_a = math.sin(angle_rad)
x1,y1,x2,y2= 256,408,470,436

cx = (x1 + x2) / 2
print("cx:",cx)
cy = (y1 + y2) / 2
print("cy:",cy)

# Translate points to origin
x1 -= cx
y1 -= cy
x2 -= cx
y2 -= cy

# Rotate points
new_x1 = x1 * cos_a - y1 * sin_a
new_y1 = x1 * sin_a + y1 * cos_a
new_x2 = x2 * cos_a - y2 * sin_a
new_y2 = x2 * sin_a + y2 * cos_a

# Translate points back
new_x1 += cx
new_y1 += cy
new_x2 += cx
new_y2 += cy

draw.rectangle([new_x1,new_y1,new_x2,new_y2],outline="red",width=2)
rotate.save(rotate_image)
rotate.show()
```
the rectangle showing a wrong.Please  give me another formula or suggestion.This is not a leukemia cell detection model.I have attached sample photo below
![Aadhaar_letter_large](https://github.com/ultralytics/ultralytics/assets/127183096/a238804d-baf8-4766-bfcb-afa9f277b808)
like that i need to give 360 degree images.my model does not work for 90 to 270 degree.so i need to do automation  for 360 degree

**Reply by pderrenger** [reply]
@Nithishrish23 it seems like you're trying to rotate a bounding box and then draw the rotated box on the rotated image. The issue with your current approach is that you're rotating the image first and then trying to apply the rotation to the bounding box coordinates, which will not align correctly because the image has already been transformed.

To fix this, you should rotate the bounding box coordinates first and then rotate the image. Here's a revised version of your code that should work:

```python
from PIL import Image, ImageDraw
import math

image_dir = r"1.jpg"
image = Image.open(image_dir)
rotate_image = r"1_rotated.jpg"

angle = 1  # Angle by which you want to rotate the image and bounding box

# Original bounding box coordinates
x1, y1, x2, y2 = 256, 408, 470, 436

# Calculate the center of the original image
cx_original = (x1 + x2) / 2
cy_original = (y1 + y2) / 2

# Calculate the center of the rotated image
w, h = image.size
cx_rotated = w / 2
cy_rotated = h / 2

# Calculate the angle in radians
angle_rad = math.radians(angle)

# Calculate the rotation matrix
cos_a = math.cos(angle_rad)
sin_a = math.sin(angle_rad)

# Function to rotate a point around a center
def rotate_point(x, y, cx, cy, cos_a, sin_a):
    x -= cx
    y -= cy
    new_x = x * cos_a - y * sin_a
    new_y = x * sin_a + y * cos_a
    new_x += cx
    new_y += cy
    return new_x, new_y

# Rotate the bounding box coordinates
new_x1, new_y1 = rotate_point(x1, y1, cx_original, cy_original, cos_a, sin_a)
new_x2, new_y2 = rotate_point(x2, y2, cx_original, cy_original, cos_a, sin_a)

# Rotate the image
rotate = image.rotate(angle, resample=Image.Resampling.BICUBIC, expand=True)

# Draw the rotated bounding box on the rotated image
draw = ImageDraw.Draw(rotate)
draw.rectangle([new_x1, new_y1, new_x2, new_y2], outline="red", width=2)

# Save and show the result
rotate.save(rotate_image)
rotate.show()
```

This code rotates the points first and then rotates the image, so the bounding box should align correctly with the rotated image.

Regarding the 360-degree rotation for your model, you can create a loop to generate images rotated at different angles and then use those images to train your model. This way, your model will learn to detect objects at various orientations, which should improve its performance for angles between 90 to 270 degrees.

**Reply by Nithishrish23** [reply]
cx_rotated = w / 2
cy_rotated = h / 2


where i want use

**Reply by pderrenger** [reply]
@Nithishrish23 the `cx_rotated` and `cy_rotated` represent the center coordinates of the rotated image. In the provided code snippet, they are not being used, but if you need to translate the rotated bounding box to align with the center of the rotated image, you would use these variables. If your bounding box is correctly positioned without this translation, you can omit these lines.

**Comment by Nithishrish23** [comment]
@glenn-jocher  @pderrenger 
Are u able to explain how model train.i have an small qurey about did not mentioned documentation clearly

- [ ]  Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
- [ ]       1/688      2.24G     0.2535     0.1879     0.7963         36        640: 100%|██████████| 688/688 [12:51<00:00,  1.12s/it]
- [ ]                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 344/344 [09:16<00:00,  1.62s/it]
- [ ]                    all      11002      21912      0.936       0.86      0.933       0.89

                   
normal epoch formula = no of images/no of batches. Here 688/688 has completed in 1st itself. Does it enough or i need to proceed 688 times            
688/688 what is the meaning of this sir or mam

GPU Memory: 2.24G: The GPU memory consumption during this epoch.
Box Loss: 0.2535, Class Loss: 0.1879, DFL Loss: 0.7963: These are the losses computed during training for bounding boxes, class predictions, and other specified loss functions.
Instances: 36: The number of instances processed in this epoch.
Image Size: 640: The size of the input images during training.
Training: 100%|██████████| 688/688 [12:51<00:00, 1.12s/it]: Indicates the progress of training. It processed 688 batches in this epoch, taking a total of 12 minutes and 51 seconds, with an average of 1.12 seconds per batch.
The second part indicates the evaluation on the validation set:

Class Evaluation: 100%|██████████| 344/344 [09:16<00:00, 1.62s/it]: The model evaluated on 344 batches of the validation set, taking 9 minutes and 16 seconds.
Overall Metrics: Shows the aggregated metrics over the entire validation set.
Total Images: 11002, Total Instances: 21912: The number of images and instances in the validation set.
Box Precision: 0.936, Box Recall: 0.86: Precision and recall for bounding box predictions.
mAP50: 0.933, mAP50-95: 0.89: Mean Average Precision (mAP) at different IoU thresholds.
wheather training process completed one epoch, processing 688 batches. Whether 688 epochs are needed depends on the convergence of the model and the task's requirements. The metrics provided can help assess the model's performance.

#Help me to sort it out

i my preditiction model it is overlapping and predict unwanted not giving exact accuracy i need 99 percent right now 91 percent.Extra bounding box are detecting.Is any possibility to make 100 Percentage

**Reply by glenn-jocher** [reply]
@Nithishrish23 hello! Let's clarify your queries about the YOLOv8 training process and address the issue with your prediction model.

### Training Process Explanation
- **688/688**: This notation indicates that one epoch has completed, where all 688 batches of your dataset were processed. It does not mean 688 epochs; it's just one epoch. You should continue training for more epochs as defined in your training setup (e.g., 688 epochs if that's what you've set) or until your model converges and stops improving.

### Prediction Model Issue
- **Overlapping and Extra Bounding Boxes**: This is a common issue that can often be improved by adjusting the confidence threshold and Non-Maximum Suppression (NMS) settings. Here's a quick tip on how you might adjust these parameters:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Set higher confidence threshold and adjust NMS
conf = 0.5  # Only detect objects with confidence >= 50%
iou = 0.5   # Adjust IoU threshold for NMS

# Run prediction
results = model.predict('path/to/image.jpg', conf=conf, iou=iou)
results.show()  # Display the predictions
```

- **Achieving 100% Accuracy**: It's challenging to achieve 100% accuracy in real-world scenarios, especially in complex tasks like object detection. Focus on improving the model incrementally and consider gathering more diverse training data or further tuning your model's hyperparameters.

Feel free to continue training, monitor the validation metrics, and adjust based on the performance insights you gather. If you have more specific questions or need further assistance, don't hesitate to ask! 🚀
